File: generated_code/GPT4/Java/051_SplayTree.java
public class SplayTree {
    // Node class representing each node in the Splay Tree
    private class Node {
        int key;
        Node left, right, parent;

        Node(int key) {
            this.key = key;
        }
    }

    private Node root;

    // Public method to insert a key into the Splay Tree
    public void insert(int key) {
        if (root == null) {
            root = new Node(key);
            return;
        }

        Node node = root;
        Node parent = null;
        while (node != null) {
            parent = node;
            if (key < node.key) {
                node = node.left;
            } else if (key > node.key) {
                node = node.right;
            } else {
                // Key already exists, splay and return
                splay(node);
                return;
            }
        }

        Node newNode = new Node(key);
        newNode.parent = parent;
        if (key < parent.key) {
            parent.left = newNode;
        } else {
            parent.right = newNode;
        }
        splay(newNode);
    }

    // Public method to delete a key from the Splay Tree
    public void delete(int key) {
        Node node = findNode(key);
        if (node == null) {
            return; // Key not found, do nothing
        }
        splay(node);

        if (node.left == null) {
            replace(node, node.right);
        } else if (node.right == null) {
            replace(node, node.left);
        } else {
            Node min = subtreeMin(node.right);
            if (min.parent != node) {
                replace(min, min.right);
                min.right = node.right;
                if (min.right != null) {
                    min.right.parent = min;
                }
            }
            replace(node, min);
            min.left = node.left;
            if (min.left != null) {
                min.left.parent = min;
            }
        }
        node.left = node.right = node.parent = null; // Help GC
    }

    // Public method to search for a key in the Splay Tree
    // Returns the key if found, or null if not found
    public Integer search(int key) {
        Node node = findNode(key);
        if (node != null) {
            splay(node);
            return node.key;
        }
        return null;
    }

    // Helper: Find the node with the specified key
    private Node findNode(int key) {
        Node node = root;
        while (node != null) {
            if (key < node.key) {
                node = node.left;
            } else if (key > node.key) {
                node = node.right;
            } else {
                return node;
            }
        }
        return null;
    }

    // Helper: Replace node u with node v in the tree
    private void replace(Node u, Node v) {
        if (u.parent == null) {
            root = v;
        } else if (u == u.parent.left) {
            u.parent.left = v;
        } else {
            u.parent.right = v;
        }
        if (v != null) {
            v.parent = u.parent;
        }
    }

    // Helper: Find the minimum node in a subtree
    private Node subtreeMin(Node node) {
        while (node.left != null) {
            node = node.left;
        }
        return node;
    }

    // Splaying operation: brings the given node to the root
    private void splay(Node x) {
        if (x == null) return;
        while (x.parent != null) {
            if (x.parent.parent == null) {
                // Zig case
                if (x == x.parent.left) {
                    rightRotate(x.parent);
                } else {
                    leftRotate(x.parent);
                }
            } else if (x == x.parent.left && x.parent == x.parent.parent.left) {
                // Zig-Zig case (left-left)
                rightRotate(x.parent.parent);
                rightRotate(x.parent);
            } else if (x == x.parent.right && x.parent == x.parent.parent.right) {
                // Zig-Zig case (right-right)
                leftRotate(x.parent.parent);
                leftRotate(x.parent);
            } else if (x == x.parent.right && x.parent == x.parent.parent.left) {
                // Zig-Zag case (left-right)
                leftRotate(x.parent);
                rightRotate(x.parent);
            } else {
                // Zig-Zag case (right-left)
                rightRotate(x.parent);
                leftRotate(x.parent);
            }
        }
    }

    // Left rotation at node x
    private void leftRotate(Node x) {
        Node y = x.right;
        if (y == null) return;
        x.right = y.left;
        if (y.left != null) {
            y.left.parent = x;
        }
        y.parent = x.parent;
        if (x.parent == null) {
            root = y;
        } else if (x == x.parent.left) {
            x.parent.left = y;
        } else {
            x.parent.right = y;
        }
        y.left = x;
        x.parent = y;
    }

    // Right rotation at node x
    private void rightRotate(Node x) {
        Node y = x.left;
        if (y == null) return;
        x.left = y.right;
        if (y.right != null) {
            y.right.parent = x;
        }
        y.parent = x.parent;
        if (x.parent == null) {
            root = y;
        } else if (x == x.parent.right) {
            x.parent.right = y;
        } else {
            x.parent.left = y;
        }
        y.right = x;
        x.parent = y;
    }

    // Optional: For testing, in-order traversal
    /*
    public void inorder() {
        inorder(root);
        System.out.println();
    }

    private void inorder(Node node) {
        if (node != null) {
            inorder(node.left);
            System.out.print(node.key + " ");
            inorder(node.right);
        }
    }
    */
}

File: generated_code/GPT4/Python/051_SplayTree.py
class SplayTree:
    class Node:
        __slots__ = 'key', 'left', 'right', 'parent'
        def __init__(self, key, parent=None):
            self.key = key
            self.left = None
            self.right = None
            self.parent = parent

    def __init__(self):
        self.root = None

    # Helper: Rotate x up
    def _rotate(self, x):
        p = x.parent
        if p is None:
            return  # x is root
        g = p.parent

        if p.left == x:
            # Right rotation
            p.left = x.right
            if x.right:
                x.right.parent = p
            x.right = p
        else:
            # Left rotation
            p.right = x.left
            if x.left:
                x.left.parent = p
            x.left = p

        p.parent = x
        x.parent = g

        if g:
            if g.left == p:
                g.left = x
            else:
                g.right = x
        else:
            self.root = x

    # Splaying x to root (or closest possible if x is None)
    def _splay(self, x):
        if x is None:
            return
        while x.parent:
            p = x.parent
            g = p.parent
            if not g:
                self._rotate(x)
            elif (g.left == p) == (p.left == x):  # Zig-zig
                self._rotate(p)
                self._rotate(x)
            else:  # Zig-zag
                self._rotate(x)
                self._rotate(x)

    # Search for key, splay the found node (or last accessed node) to root
    def search(self, key):
        node = self.root
        last = None
        while node:
            last = node
            if key == node.key:
                self._splay(node)
                return True
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        if last:
            self._splay(last)
        return False

    # Insert key (if not present), splay new node to root
    def insert(self, key):
        if not self.root:
            self.root = self.Node(key)
            return
        node = self.root
        parent = None
        while node:
            parent = node
            if key == node.key:
                self._splay(node)
                return  # Already exists, do nothing
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        newnode = self.Node(key, parent)
        if key < parent.key:
            parent.left = newnode
        else:
            parent.right = newnode
        self._splay(newnode)

    # Find node with key
    def _find(self, key):
        node = self.root
        last = None
        while node:
            last = node
            if key == node.key:
                return node
            elif key < node.key:
                node = node.left
            else:
                node = node.right
        return None

    # Join two trees: left's all keys < right's all keys
    def _join(self, left, right):
        if not left:
            return right
        if not right:
            return left
        # Max of left
        node = left
        while node.right:
            node = node.right
        self._splay_node(node, left)
        node.right = right
        if right:
            right.parent = node
        return node

    # Splay node x, given a custom root
    def _splay_node(self, x, root):
        while x.parent:
            p = x.parent
            g = p.parent
            if not g:
                self._rotate_custom(x, root)
            elif (g.left == p) == (p.left == x):  # Zig-zig
                self._rotate_custom(p, root)
                self._rotate_custom(x, root)
            else:  # Zig-zag
                self._rotate_custom(x, root)
                self._rotate_custom(x, root)

    def _rotate_custom(self, x, root):
        p = x.parent
        if p is None:
            return
        g = p.parent

        if p.left == x:
            p.left = x.right
            if x.right:
                x.right.parent = p
            x.right = p
        else:
            p.right = x.left
            if x.left:
                x.left.parent = p
            x.left = p

        p.parent = x
        x.parent = g

        if g:
            if g.left == p:
                g.left = x
            else:
                g.right = x
        # else: do not update self.root

    # Delete key if present. Splay the parent of deleted node (or root) to root.
    def delete(self, key):
        node = self.root
        parent = None
        while node:
            if key == node.key:
                break
            parent = node
            if key < node.key:
                node = node.left
            else:
                node = node.right
        if not node:
            if parent:
                self._splay(parent)
            return  # Not found
        self._splay(node)
        # Now node is root
        left = node.left
        right = node.right
        if left:
            left.parent = None
        if right:
            right.parent = None
        # Join left and right as new tree
        self.root = self._join(left, right)

    # For debugging: inorder traversal
    def _inorder(self, node, res):
        if not node:
            return
        self._inorder(node.left, res)
        res.append(node.key)
        self._inorder(node.right, res)

    # Return sorted list of keys
    def keys(self):
        res = []
        self._inorder(self.root, res)
        return res